WordDistanceInSpark.scala

```diff
@@ -1 +1,241 @@
+    // I'm not sure whether defining a custom accumulator would help for making it less ugly 
+    // and for making it more performant.
+     
+    val reverseOrderedSameWordIntervals:List[SameWordInterval] = convertHitsToSameWordIntervals(orderedHits)
+    // the fact that I do a fold which is an action instead of a transform, 
+    // is bumping me out of RDDs into regular scala-land at this point anyway.
+    // At this point we have had another huge data shrink from skipping consecutive hits of the same word,
+    // and it is already late, so I'm going to do the last part in pure scala
+    
+    // println("""***** """+reverseOrderedSameWordIntervals.take(100).toList.mkString(", "))
+    
+    // if one of the words does not exist, there will only be one element in this list
+    if (reverseOrderedSameWordIntervals.size == 1) {
+      None 
+    } else {
+      Some(findMinimumDistanceBetweenAlternatingIntervals(reverseOrderedSameWordIntervals))
+    }
+  }
+   
+  def convertHitsToSameWordIntervals(orderedHits:RDD[Hit]):List[SameWordInterval] = {
+    // a hit becomes an interval with same start as end
+    orderedHits.map(hit => List(((hit._1,hit._1), hit._2))).
+      fold(List[SameWordInterval]())(WordDistanceInSpark.foldIntoExistingSameWordIntervals _)
+  }
+  
+  def foldIntoExistingSameWordIntervals(acc:List[SameWordInterval], extra:List[SameWordInterval]): List[SameWordInterval] = {
+    // warning: car/cdr soup ahead. Which is why this part is unittested and gets a post-condition.
+    def wordAt(swi: SameWordInterval):String = swi._2
+    
+    //println("* Adding "+extra +" into "+acc)
+    
+    var shouldMerge = false
+    val result = if (acc.isEmpty) { 
+      extra
+    } else {
+        val endOfAcc = acc.head
+        val beginOfExtra = extra.last
+        shouldMerge = wordAt(endOfAcc)==wordAt(beginOfExtra)
+        if (shouldMerge) {
+          val mergedLink = ((acc.head._1._1, extra.last._1._2), wordAt(endOfAcc))
+          extra.dropRight(1):::(mergedLink::acc.tail)
+        } else {
+          extra:::acc
+        }
+    }
+
+    val expectedSize = acc.size + extra.size - (if (shouldMerge) 1 else 0)
+    if (result.size != expectedSize) throw new Exception("Programming error in merging "+extra +" into acc "+acc+": wrong size of result "+result)
+    result
+  }
+  
+  def findMinimumDistanceBetweenAlternatingIntervals(reverseOrderedSameWordIntervals:List[SameWordInterval]):Long = {
+    // in pure scala I get sliding, in spark this seems to be part of the MLLib. I have not investigated further.
+    val distances = reverseOrderedSameWordIntervals.sliding(2).map {
+        case laterInterval::precedingInterval::Nil => calcDist(precedingInterval, laterInterval)
+      }
+    distances.min
+  }
+
+  def calcDist(first:SameWordInterval, second:SameWordInterval):Long = calcDist(first._1._2, second._1._1)
+  
+  def calcDist(first:Long, last:Long):Long = last-first-1
```

Comment:
Do we need two overloads of this method?  Probably same comment as above, do we need a separate method for a simple function?